File: server/app/models/device.py
from sqlalchemy import Column, BigInteger, String, Enum, DateTime, JSON
from sqlalchemy.sql import func
from ..core.database import Base


class Device(Base):
    __tablename__ = "devices"

    id = Column(BigInteger, primary_key=True, autoincrement=True)
    device_code = Column(String(64), unique=True, nullable=False, index=True)
    name = Column(String(128))
    location = Column(String(256))
    firmware_version = Column(String(32))
    status = Column(Enum("online", "offline", "busy", "error", name="device_status"), default="offline")
    last_heartbeat = Column(DateTime)
    last_upload = Column(DateTime)
    metadata_ = Column("metadata", JSON)
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())

File: web/src/api/device.ts
import request from './request'

export interface Device {
  id: number
  device_code: string
  name: string | null
  location: string | null
  firmware_version: string | null
  status: 'online' | 'offline' | 'busy' | 'error'
  last_heartbeat: string | null
  last_upload: string | null
  created_at: string
}

export interface DeviceCreate {
  device_code: string
  name?: string
  location?: string
  firmware_version?: string
}

export const deviceApi = {
  list(params?: { status?: string; skip?: number; limit?: number }) {
    return request.get<Device[]>('/device/list', { params })
  },

  get(id: number) {
    return request.get<Device>(`/device/${id}`)
  },

  create(data: DeviceCreate) {
    return request.post<Device>('/device/register', data)
  },

  delete(id: number) {
    return request.delete(`/device/${id}`)
  },

  heartbeat(deviceCode: string, status: string = 'online') {
    return request.post('/device/heartbeat', { device_code: deviceCode, status })
  },
}
